packages/adapter/src/lib/adapter/utils.ts

```diff
@@ -1,6 +1,16 @@
 import { SYSTEM_ADMIN_USER } from './constants';
 import { tools, EXIT_CODES } from '@iobroker/js-controller-common';
 
+type OpaqueString<T extends string> = string & {
+    __type: T;
+};
+
+export type ID = OpaqueString<'ID'>;
+
+type Callback = (...args: any[]) => void | Promise<void>;
```

Comment:
It could make sense to type this and the assert functions as generics, so you can pass the argument types and not introduce `any`s this way.

Reply:
I am not sure how I can infer the types here because of the rest params. We have a not known amount of args here which can also differ in type.

Reply:
Example:
```ts
type Callback<T extends any[] = any[]> = (...args: T) => void | Promise<void>;
type OptionalCallback<T extends any[] = any[]> = undefined | Callback<T>;

static assertsOptionalCallback<T extends any[] = any[]>(value: unknown, name: string): asserts value is OptionalCallback<T> {
  // ... implementation as it is now
}
```
and then at the usage site:

```ts
// No generic argument, everything like it is now:
assertsOptionalCallback(cb);
// cb now has the type ( (...args: T) => void | Promise<void> ) | undefined

// Or if we know the arguments it should have
assertsOptionalCallback<[param1: number, param2: string]>(cb);
// cb now has the type ( (param1: number, param2: string) => void | Promise<void> ) | undefined
```
Note that this doens't catch the case that the callback has incorrect parameters, but at least it makes it somewhat easier to work with the callback, or catch errors at the calling site.

Reply:
I am currently quite unsure about this, beside the fact that I need pass all types to the callbacks manually. 

We are just validating in the method, that it is a function, then it is passed to the actual implementation method, which then has correct callback type again. see e.g. here https://github.com/ioBroker/ioBroker.js-controller/blob/12fdbae1f6b15d7626aef56c51256cacfd5f97d7/packages/adapter/src/lib/adapter/adapter.ts#L1253-L1261

with https://github.com/ioBroker/ioBroker.js-controller/blob/12fdbae1f6b15d7626aef56c51256cacfd5f97d7/packages/adapter/src/lib/_Types.ts#L99